Internal: Class usage count locator [ED-19152]- for remove

MuliElementor opened the pull request:
## PR Checklist
<!-- 
Please check if your PR fulfills the following requirements:
**Filling out the template is required.** Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
 -->
- [ ] The commit message follows our guidelines:  https://github.com/elementor/elementor/blob/master/.github/CONTRIBUTING.md


## PR Type
What kind of change does this PR introduce?
<!-- Please check the one that applies to this PR using "x" with no spaces eg: [x]. -->
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Summary
Adding CSS Classes Usage API with optional enrichment


*

## Description
This PR introduces a new REST API endpoint under elementor/v1/global-classes-usage that returns detailed usage statistics of global CSS classes across Elementor documents.

Supports optional enrichment via with_page_info query param.

When with_page_info=true, each usage entry includes the associated post/page title.

Added flag-aware logic to avoid unnecessary get_the_title() calls when not needed.

Follows Elementor's internal REST conventions (route_wrapper, Response_Builder, permission_callback, and args).

Ensures manage_options permission guard is respected.

Also updated the JS client (usage()) to support toggling the flag via request param.
*

## Test instructions
This PR can be tested by following these steps:

*

## Quality assurance

- [ ] I have tested this code to the best of my abilities
- [ ] I have added unittests to verify the code works as intended
- [ ] Docs have been added / updated (for bug fixes / features)

Fixes #

<!--start_gitstream_placeholder-->
### ✨ PR Description
Purpose: Add global CSS class usage tracking and visualization with a class locator UI to improve management of global classes.
Main changes:
- Implemented backend API endpoint for tracking global class usage across documents
- Created CSS class usage data model with detailed usage statistics per document
- Added locator UI component showing where each global class is being used

_Generated by LinearB AI and added by gitStream._
<sub>AI-generated content may contain inaccuracies. Please verify before using. **[We'd love your feedback!](mailto:product@linearb.io)** 🚀</sub>
<!--end_gitstream_placeholder-->

mulituriel-beep committed:
create usage prop in rest response

mulituriel-beep committed:
multiple changes

mulituriel-beep committed:
add page title

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/elementor/elementor?pullRequest=31644) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/elementor/elementor?pullRequest=31644) before we can accept your contribution.<br/><hr/>**mulituriel_tikal** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/elementor/elementor?pullRequest=31644) it.</sub>

mulituriel-beep committed:
update

mulituriel-beep committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

# Conflicts:
#	package-lock.json

mulituriel-beep committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
Skip processing templates in applied-global-classes-usage script.

mulituriel-beep committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
Add tests and components for tracking CSS class usage across project

mulituriel-beep committed:
Refactor and optimize CSS class usage logic

Removed redundant tests and mocks while consolidating hook logic for `useCssClassUsage`. Simplified related components and improved testing structure. Minor updates to query keys and default usage handling.

mulituriel-beep committed:
Refactor and optimize CSS class usage logic

Removed redundant tests and mocks while consolidating hook logic for `useCssClassUsage`. Simplified related components and improved testing structure. Minor updates to query keys and default usage handling.

mulituriel-beep committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
update

mulituriel-beep committed:
update

mulituriel-beep committed:
update

MuliElementor committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
Merge remote-tracking branch 'origin/ED-19152-class-usage-count-locator' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
update

mulituriel-beep committed:
update

mulituriel-beep committed:
update

mulituriel-beep committed:
update

mulituriel-beep committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
update tests

MuliElementor committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

MuliElementor committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
Merge remote-tracking branch 'origin/ED-19152-class-usage-count-locator' into ED-19152-class-usage-count-locator

MuliElementor committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
update

MuliElementor committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
update

mulituriel-beep committed:
update

mulituriel-beep committed:
update

mulituriel-beep committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
fix tests

mulituriel-beep committed:
update tests

mulituriel-beep committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

# Conflicts:
#	modules/global-classes/global-classes-css.php

mulituriel-beep committed:
update type

MuliElementor committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
update type

mulituriel-beep committed:
Merge remote-tracking branch 'origin/ED-19152-class-usage-count-locator' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
update

mulituriel-beep committed:
update test

MuliElementor committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
remove test

mulituriel-beep committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
Refactor and enhance CSS class usage tracking.

Replaced legacy class usage logic with new modular components, including `Css_Class_Usage` and `Document_Usage`. Introduced detailed usage tracking and updated related API calls and tests for improved maintainability and accuracy.

mulituriel-beep committed:
Merge remote-tracking branch 'origin/ED-19152-class-usage-count-locator' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
Refactor tooltip and button for improved disabled state handling

Replaced static tooltip text with dynamic infotip content for better UX. Introduced a styled `CustomIconButton` to improve hover behavior on disabled buttons, providing a more intuitive and interactive experience.

mulituriel-beep committed:
Refactor and enhance `CssClassUsageTrigger` with callback support.

Added an `onClick` callback to `CssClassUsageTrigger` to handle ID updates. Refactored related logic to improve readability and maintainability, while simplifying the handling of popover states in `class-item`.

mulituriel-beep committed:
update

mulituriel-beep committed:
update tooltip for "not is use"

mulituriel-beep committed:
Update tooltip to use dynamic text for location count

Replaced static tooltip text with dynamic content that adjusts based on the number of locations. This ensures better localization and improves user clarity by handling singular and plural cases properly.

mulituriel-beep committed:
Enhance class usage tracking by adding document type support

mulituriel-beep committed:
Add icons to CssClassUsagePopover items

Introduce an `iconMap` to map item types to corresponding icons, improving visual distinction in the CssClassUsagePopover component. Updated the item structure to include type-based icons and adjusted styles for better spacing. Extended the `CssClassUsageContent` type to include the `type` property necessary for this enhancement.

mulituriel-beep committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
Add support for opening documents in a new tab

Refactor `useNavigateToDocument` to accept an `openNewTab` option. When enabled, documents can now be opened in a new browser tab while preserving the existing navigation logic.

mulituriel-beep committed:
Adjust horizontal popover offset in CSS class usage trigger.

The horizontal offset was updated from -20 to -5 to improve the positioning of the popover. This ensures better alignment and visual consistency in the UI.

mulituriel-beep committed:
Add navigation to document in CSS Class Usage Popover

Introduce an external link button for navigating to documents directly from the CSS Class Usage Popover. Updated icon mappings and incorporated new header/footer icons for better template identification.

mulituriel-beep committed:
Refactor and enhance icon mapping in CssClassUsagePopover

Replaced `iconMap` with a more robust `iconMapper` to include labels for icons. Added a `Tooltip` to display icon labels, improving accessibility and user experience. Adjusted item structure to incorporate `docType` for better context handling.

mulituriel-beep committed:
Enhance global CSS class usage tracking with detailed reporting and exclusion of specific document types

mulituriel-beep committed:
Merge branch 'refs/heads/main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
Enhance global CSS class usage tracking with detailed reporting and exclusion of specific document types

mulituriel-beep committed:
Refactor content type handling with a centralized enum

Introduce `ContentType` enum to standardize handling for content types like Post, Page, Popup, Header, and Footer. Update related components, types, and tests to use `ContentType` for improved type safety and consistency. Adjust PopoverBody width and fix tests with appropriate type definitions.

mulituriel-beep committed:
Add tooltip to external link button and improve accessibility in usage trigger

mulituriel-beep committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
Fix document type retrieval logic to handle empty values more gracefully

mulituriel-beep committed:
update

mulituriel-beep committed:
Refactor css-class-usage popover for improved UI and usability.

Replaces PageTypeIcon with PagesIcon, enhances tooltip handling, and introduces `EllipsisWithTooltip` for better text truncation. Updates hover effects, improves layout with `Stack` components, and refines pointer behavior for a cleaner, more user-friendly interface.

mulituriel-beep committed:
fix tests

mulituriel-beep committed:
Refactor navigation hooks and add new tab support

Simplified `useNavigateToDocument` by extracting new tab functionality into a separate `useOpenDocumentInNewTab` hook. Introduced a shared `getUpdateUrl` utility for URL creation to improve reusability and maintainability. Added unit tests to verify the behavior of the new hook.

mulituriel-beep committed:
Refactor CSS class usage popover for improved type safety

Replaced `ContentType` enum with a union type for better type safety and maintainability. Updated navigation function to use `__useOpenDocumentInNewTab` for clarity and simplified related logic. Adjusted test cases to align with the new structure and ensure consistent functionality.

mulituriel-beep committed:
fix tests

mulituriel-beep committed:
fix

mulituriel-beep committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
בם

MuliElementor committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
update

MuliElementor committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

MuliElementor committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
update

mulituriel-beep committed:
update

mulituriel-beep committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
update

MuliElementor committed:
Update modules/global-classes/global-classes-rest-api.php

Co-authored-by: gitstream-cm[bot] <111687743+gitstream-cm[bot]@users.noreply.github.com>

mulituriel-beep committed:
update

mulituriel-beep committed:
revert

mulituriel-beep committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
fix ui

mulituriel-beep committed:
update tests

mulituriel-beep committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

Nevoss commented on packages/packages/core/editor-global-classes/src/components/css-class-usage/hooks/use-css-class-usage.ts:
```diff
@@ -0,0 +1,13 @@
+	return useQuery( {
+		queryKey: [ QUERY_KEY ],
+		queryFn: fetchCssClassUsage,
+		refetchOnMount: false,
```
Why disable `refetchOnMount`?

Nevoss commented on packages/packages/core/editor-global-classes/src/components/css-class-usage/components/css-class-usage-popover.tsx:
```diff
@@ -0,0 +1,160 @@
+
+type VirtualItem = VirtualizedItem< 'item', string > & { docType: ContentType };
+
+const iconMapper: Record< ContentType, { label: string; icon: React.ReactElement } > = {
```
There are a lot of post types that are generated by other plugins and Elementor Pro as well. consider something like `unknow` type, and also maybe we should allow extending this record.

MuliElementor replied:
i added fallback, for icon and label

Nevoss commented on packages/packages/core/editor-global-classes/src/components/css-class-usage/components/css-class-usage-popover.tsx:
```diff
@@ -0,0 +1,160 @@
+	const { data: classUsage } = useCssClassUsageByID( cssClassID );
+	const onNavigate = useOpenDocumentInNewTab();
+
+	const items: VirtualItem[] =
```
Please consider better naming. What is the name of this entity? Also, `VirtualItem` is a little bit confusing.

MuliElementor replied:
changed to "cssClassUsageRecords"

Nevoss commented on packages/packages/core/editor-global-classes/src/components/css-class-usage/components/css-class-usage-popover.tsx:
```diff
@@ -0,0 +1,160 @@
+	);
+};
+
+const StyledCssClassUsageItem = styled( MenuList )( ( { theme } ) => ( {
```
Talk with @IshayMaya: Are these our best practices? We try to avoid creating custom styles, am I wrong?

Nevoss commented on modules/global-classes/usage/document-usage.php:
```diff
@@ -0,0 +1,101 @@
+	 */
+	protected function get_all_global_class_ids(): array {
+		return Global_Classes_Repository::make()
+										->all()
+										->get_items()
+										->filter( fn( $item ) => ! empty( $item['id'] ?? null ) )
+										->keys()
+										->all();
```
Is global class without id is possible?

Nevoss commented on modules/global-classes/usage/document-usage.php:
```diff
@@ -0,0 +1,101 @@
+	 */
+	public function __construct( ElementorDocument $document ) {
+		$this->document      = $document;
+		$this->elements_data = $document->get_elements_raw_data() ?? [];
```
Consider calling this query when needed and not in the constructor

Nevoss commented on modules/global-classes/usage/document-usage.php:
```diff
@@ -0,0 +1,101 @@
+	 */
+	public function analyze(): void {
+		$page_id       = $this->document->get_main_id();
+		$page_title    = get_the_title( $page_id );
```
You can get the title from the post, `$this-document->post->post_title`

Nevoss reviewed (dismissed):
Please check why there is a need for a change in `package-lock`? We don't want any surprises in terms of package versions.

And please take a look at @Omerisra6 comments, and fix what's needed.

mulituriel-beep committed:
update
fix issue

mulituriel-beep committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

# Conflicts:
#	package-lock.json

MuliElementor commented on packages/packages/core/editor-global-classes/src/components/class-manager/class-item.tsx:
```diff
@@ -159,22 +164,31 @@ export const ClassItem = ( {
-		}
-	}
-`
+	( { showActions, showSortIndicator } ) => `
```
we can remote one of them, it was before v3.30 was released 
the V2 just a mark for the "new" design.
i will remove it and change it permanently

mulituriel-beep committed:
update route variable
remove unused params

mulituriel-beep committed:
remove unused params

mulituriel-beep committed:
feat: Implement CSS class usage service and hooks

- Added `fetchCssClassUsage` function to retrieve CSS class usage data from the API.
- Created `useCssClassUsage` hook for querying CSS class usage data.
- Introduced `useCssClassUsageByID` hook to access specific class usage by ID.
- Added `usePrefetchCssClassUsage` hook for prefetching CSS class usage data.
- Removed obsolete hooks and index files related to CSS class usage.

mulituriel-beep committed:
Refactor Document_Usage class for improved readability and constructor logic

- Reformatted class and method declarations to PSR-12 style
- Removed unused `elements_data` property from the class
- Updated constructor to only assign the document object, removing direct element data assignment
- In `analyze` method, replaced direct call to `get_the_title()` with accessing the post title via the document object
- Minor code style adjustments throughout for consistency

MuliElementor committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
fix llint

mulituriel-beep committed:
update

mulituriel-beep committed:
update

mulituriel-beep committed:
fix: sanitize Icon props to prevent non-boolean attribute warnings and update related tests

- Refactor attribute handling in css-class-usage-service and class-item components to ensure only valid HTML props are passed to Icon
- Prevent passing boolean values to Icon attributes, resolving React warnings about non-boolean attributes
- Update all related tests to reflect changes in prop handling and eliminate warnings

MuliElementor commented on packages/packages/libs/editor-ui/src/components/popover/header.tsx:
```diff
@@ -5,7 +5,7 @@ import { CloseButton, Stack, Typography } from '@elementor/ui';
 
 type PopoverHeaderProps = {
-	title: string;
+	title: React.ReactNode;
```
it will break the api, i prefer to keep it like this

MuliElementor commented on tests/phpunit/elementor/modules/global-classes/test-css-class-usage.php:
```diff
@@ -0,0 +1,79 @@
+use Elementor\Modules\GlobalClasses\Usage\Css_Class_Usage;
+use PHPUnit\Framework\TestCase;
+
+class Test_Css_Class_Usage extends TestCase {
```
possible

MuliElementor commented on packages/packages/core/editor-global-classes/src/components/css-class-usage/index.ts:
```diff
@@ -0,0 +1,3 @@
+export * from './types';
+export * from './hooks';
+export * from './components';
```
move theme

MuliElementor commented on packages/packages/core/editor-global-classes/src/components/css-class-usage/__tests__/use-css-class-usage-by-id.test.tsx:
```diff
@@ -0,0 +1,69 @@
+	} );
+
+	it( 'returns usage data for a valid ID', async () => {
+		mockedFetchCssClassUsage.mockResolvedValue( {
```
done

mulituriel-beep committed:
fix mock type

MuliElementor commented on packages/packages/core/editor-global-classes/src/components/css-class-usage/__tests__/css-class-usage-service.test.tsx:
```diff
@@ -0,0 +1,79 @@
+		);
+
+		it( 'should prefetch usage data and cache it', async () => {
+			( apiClient.usage as jest.Mock ).mockResolvedValue( {
```
done, expect 1

MuliElementor commented on packages/packages/core/editor-global-classes/src/components/class-manager/class-item.tsx:
```diff
@@ -103,6 +105,9 @@ export const ClassItem = ( {
 								<EllipsisWithTooltip title={ label } as={ Typography } variant="caption" />
 							) }
 						</Indicator>
+						<Box className={ 'class-item-locator' }>
```
same 
https://github.com/elementor/elementor/pull/31644#discussion_r2207479962

MuliElementor commented on modules/global-classes/usage/document-usage.php:
```diff
@@ -0,0 +1,101 @@
+										->get_items()
+										->filter( fn( $item ) => ! empty( $item['id'] ?? null ) )
+										->keys()
+										->all();
```
fixed

ronros-elementor commented on modules/global-classes/global-classes-rest-api.php:
```diff
@@ -13,7 +14,7 @@
 	const API_NAMESPACE = 'elementor/v1';
 	const API_BASE = 'global-classes';
-
+	const API_BASE_USAGE = self::API_BASE . '-usage';
```
I'd use only `usage`, as a url segment of its own -
`elementor/v1/global-classes/usage`, as "usgae" is actually a dimension/metric/attribute  of global-classes

ronros-elementor commented on modules/global-classes/usage/applied-global-classes-usage.php:
```diff
@@ -1,92 +1,126 @@
 
-		return $count_per_class;
-	}
+			$filtered_pages = array_filter(
+				$pages,
+				fn( $page_data ) => ! in_array( $page_data['type'] ?? 'unknown', $this->excluded_types, true )
+			);
```
Maybe this could be written nicer, assign the value of `$page_data['type']` to a variable and use it?

mulituriel-beep committed:
remove unused composer action

mulituriel-beep committed:
update endpoint route

mulituriel-beep committed:
update client api
remove unknown type

mulituriel-beep committed:
remove unused files

mulituriel-beep committed:
remove unused files

mulituriel-beep committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
fix

mulituriel-beep committed:
update tests

mulituriel-beep committed:
remove comment

MuliElementor commented on packages/packages/core/editor-global-classes/src/components/css-class-usage/__tests__/css-class-usage-service.test.tsx:
```diff
@@ -0,0 +1,79 @@
+
+	describe( 'usePrefetchCssClassUsage', () => {
+		const wrapper = ( { children }: { children: React.ReactNode } ) => (
+			<QueryClientProvider client={ new QueryClient() }>{ children }</QueryClientProvider>
```
it used only once
i keep it in the same file

MuliElementor committed:
Merge branch 'main' into ED-19152-class-usage-count-locator

mulituriel-beep committed:
update

mulituriel-beep committed:
revert

mulituriel-beep committed:
rerun

MuliElementor commented:
closed
merged in https://github.com/elementor/elementor/pull/31914
